Alerts section (#6)
observability-lab-cse/observability-lab-workshop: 05-alert/README.md

1. Go over usefull in-build alerts
2. Wirte custom alerts

Do you want to be the first to know when your apps and infra are acting up? Do you want to get ahead of the problems before they ruin your day (or night)? If you answered yes to any of these questions, then this workshop section is for you!

## Why do we need alerts?

In a nutshell, alerts can help you identify and troubleshoot issues before they affect your users or customers. They can also help you to align your observability goals with your business objectives, such as customer satisfaction, revenue, and growth.

## A little bit of theory

![Diagram that explains Azure Monitor alerts.](https://learn.microsoft.com/en-us/azure/azure-monitor/alerts/media/alerts-overview/alerts.png)

> Alerts are stored for 30 days and are deleted after the 30-day retention period. You can see all alert instances for all of your Azure resources on the Alerts page in the Azure portal.

Reviewer: [nit] A bit of repetition here, maybe like this: 
```suggestion
Alerts are stored for 30 days and are deleted after this retention period. You can see all alert instances for all of your Azure resources on the Alerts page in the Azure portal.
```

Author: I rephrased to 'Alerts are stored for 30 days.'

Alerts consist of a few elements, presented in the next sections.

### Action groups

**Action groups** - Collection of notification preferences defined by the user. They are the receivers of the notifications when an alert is triggered.

You can configure an action group to notify you through one or more of these methods when an alert is triggered. This allows you to respond quickly and effectively to issues in your environment.

Each action group is given a unique name and can be used by both Azure Monitor and Service Health alerts. An action group can be used across different subscriptions, allowing you to manage your alert actions centrally.

Let's [create an Action Group](https://learn.microsoft.com/en-us/azure/azure-monitor/alerts/action-groups) for Email notification!

> 1. Log in to the Azure portal.

Reviewer: [nit] Not important, but I think this step can be skipped as we can assume people are logged in at this point.

Author: True, I'll remove it

2. Navigate to the Monitor tab.
3. Under Monitor, select Alerts.
4. Click on Manage action groups.
5. Click on "+ New action group" to create a new Action Group.

> ![Action Group](.attachments/create_ag.png)

Reviewer: This and the next image don't display for me in the generated GitHub page.
![Screenshot 2023-11-23 134051](https://github.com/observability-lab-cse/observability-lab-workshop/assets/39058899/2d5b66d4-c9a8-4e00-948c-e8cfad3a5c2e)

Author: Fixed, also renamed .attachments to images to be consistent with other sections

![Action Group](.attachments/create_ag_notification.png)

Please note that Action Groups can be configured to perform various **actions** when an alert is triggered. These actions include: Email/SMS/Push/Voice, Azure Functions, Logic App, Webhook, Event Hubs etc.

But for this excersise we can skip defining actions.

> > You can test your Action group by going to **Action Groups** section in **Alerts** and using **Test** button. Check if you get a test email!

Author: I rephrased the sentences, is it better now?

### Alert conditions

> These conditions are set by the system. When an alert fires, the alert condition is set to fired. After the underlying condition that caused the alert to fire clears, the alert condition is set to resolved.

Author: Good question. don't know what I meant :D 
I simplified it

Let's create an alert condition. Feel free to create your own if you want. Here, we will go with an alert condition about AKS pods - let's set up an email notification if the number of pod in the pod lifecycle is less than 3.

![](.attachments/create_alert_rule.png)

In the actions pick the Action group that you created previously.

![](.attachments/create_alert_rule_details.png)

>  Please note that you can set up if the alert should be resolved automatically. If we uncheck that then the user needs to manually resolve the alert.

Now let's test our alert. We need to delete one of our pods.

```
kubectl get deployment

NAME                     READY   UP-TO-DATE   AVAILABLE   AGE
devices-api              0/1     1            0           3h
devices-state-manager    1/1     1            1           3h
opentelemetrycollector   1/1     1            1           3h1m

# grab the name of one of the deployments
kubectl remove deployment devices-api

# And now wait until you get your alert email
```

To deploy again the deleted service you can use make command from the project root folder:

```
# to deploy devices-api and devices-state-manager
make deploy 
```

### User response

> The response is set by the user and doesn't change until the user changes it.

Author: I rephrased

> Now you should get our email and see the alert in the Alerts window. If you click on the fired alert you can see more details.

Reviewer 2: I think a visual for this would be nice.

One thing you can do is change the User response. Let's change it to **Acknowledged** as we now saw and resolved the issue.

> ![](.attachments/fired_alert.png)

Reviewer: These images also don't render correctly.

Author: Fixed now

Once the condition changes and the alert is not longer valid (we deployed missing pods) then the Alert changes his status to **Resolved**:

![](.attachments/alert_status.png)

### Alert processing rules 

Alert processing rules allow you to apply processing on fired alerts. Alert processing rules are different from alert rules. Alert rules generate new alerts, while alert processing rules modify the fired alerts as they're being fired.

For example, you can use Alert processing rules to temporarily disable all action groups for Alerts for specific resource (e.g. during planned maintenance).

> Another example would be to set up a specific action group for all Alerts in your resource with a Critical severity, for instance you want to notify the support team immediately when critical alert occurs. 

Reviewer: ```suggestion
Another example would be to set up a specific action group for all Alerts in your resource with a Critical severity, for instance if you want to notify the support team immediately when critical alert occurs. 
```

> Alerts can be stateless and stateful.
>
> - **Stateless alerts** fire each time the condition is met, even if fired previously.
> - **Stateful alerts** fire when the rule conditions are met, and will not fire again or trigger any more actions until the conditions are resolved.
>
> What kind of alert is the one we just created?
>
> <details markdown="1">
> Stateful! [Metric Alerts are stateful](https://learn.microsoft.com/en-us/azure/azure-monitor/alerts/alerts-metric-logs#overview) - only notifying once when alert is fired and once when alert is resolved; as opposed to Log alerts, which are stateless and keep firing at every interval if the alert condition is met.

## Log based alerts

In the previous section we have created a metric alert rule using one of the predefined alerts for AKS. Let's take a look now how we can create alerts based on the logs.

> In the previous section you have created the dashboard showing different metrics. One of them was **Average processing time for last 10min**. Let's create an alert based on it!

Reviewer 2: Maybe link to the section directly
```suggestion
In section [Custom Dashboards](TBD) you have created the dashboard showing different metrics. One of them was **Average processing time for last 10min**. Let's create an alert based on it!
```

Author: done

We are going to create an alert based on a custom log. Feel free to go to your dashboard and grab the log from it or just copy it from below:

```kql
 AppTraces
 | where isnotempty(OperationId)
 | extend startTime = iif(Message startswith "Received event", TimeGenerated, datetime(null))
 | extend endTime = iif(AppRoleName == "devices-api", TimeGenerated, datetime(null))
 | partition hint.strategy=native by OperationId (
   summarize startTime = take_any(startTime), endTime = take_any(endTime) by OperationId
   )
 | where isnotempty(startTime) and isnotempty(endTime) 
 | extend delta = (endTime - startTime) / 1s
 | project startTime, delta
 | summarize avg(delta) by bin(startTime,10m)
 | sort by startTime desc
 | take 1
```

Now go ahead and click **+New Alert Rule** and create the alert using the Alert Action Group created above.

![](.attachments/create_alert_log_based_1.png)

You can **View result and edit query in Logs** and run your query:

![](.attachments/create_alert_log_based_query.png)

![](.attachments/create_alert_log_based_2.png)

>  For the learning purposes modify the processing time threshold so that the alert is triggered.

Reviewer 2: ```suggestion
For the learning purposes modify the processing time threshold so that the alert is triggered.
```

After it's done, create a few devices using swagger, then run your simulator and observe the alerts!

Go to http://DEVICES_IP:8080/swagger-ui.html and create devices.

```
# run simulator
make deploy-devices-data-simulator 
```

## Built in alerts

Let's review first the built-in alerts in Azure Portal. It's usefult to check them out as they may be useful for your solution.

> Virtual Machines, AKS and Log Analytics workspaces support [Alert rule recommendation feature]([Manage your alert rules - Azure Monitor | Microsoft Learn](https://learn.microsoft.com/en-gb/azure/azure-monitor/alerts/alerts-manage-alert-rules#enable-recommended-alert-rules-in-the-azure-portal)). After enabling it you can use predefined Alert rules and combine them with your Action groups.

Reviewer: The link here doesn't render correctly on the generated page.

Author: Fixed

![](.attachments/log_analytics_alerts.png)

![](.attachments/log_analytics_alert_rules.png)

### Smart detection alerts

[Smart detection](https://learn.microsoft.com/en-us/azure/azure-monitor/alerts/proactive-diagnostics) performs proactive analysis of the telemetry that your app sends to Application Insights. [If there’s a sudden rise in failure rates or abnormal patterns in client or server performance, you get an alert. This feature is enabled by default and operates if your application sends enough telemetry.

To look at Smart detection rules go to Application Insights -> Smart Detection.

In Alerts view you should see also alert rules created by Smart Detector:

![](.attachments/smart_detector_alerts.png)

## Service Health alerts

What's [Azure Service Health](https://learn.microsoft.com/en-gb/azure/service-health/overview)?

> The portal provides you with a customizable dashboard which tracks the health of your Azure services in the regions where you use them. In this dashboard, you can track active events like ongoing service issues, upcoming planned maintenance, or relevant health advisories. When events become inactive, they get placed in your health history for up to 90 daysService health notifications are stored in the [Azure activity log](https://learn.microsoft.com/en-gb/azure/azure-monitor/essentials/platform-logs-overview). You can use them to create your own alerts. E.g. you can create an alert that gets triggered when one of the services in your subscription is down. The alerts can be also integrated with external alerting tools, like PageDuty.

Reviewer: ```suggestion
The portal provides you with a customizable dashboard which tracks the health of your Azure services in the regions where you use them. In this dashboard, you can track active events like ongoing service issues, upcoming planned maintenance, or relevant health advisories. When events become inactive, they get placed in your health history for up to 90 days. Service health notifications are stored in the [Azure activity log](https://learn.microsoft.com/en-gb/azure/azure-monitor/essentials/platform-logs-overview). You can use them to create your own alerts. E.g. you can create an alert that gets triggered when one of the services in your subscription is down. The alerts can be also integrated with external alerting tools, like PagerDuty.
```

## Excersise

We went through a few ways how to create the alerts. Imagine that you need to create an alert that is triggered when your devices-api application is not healthy.

How would you do that? Discuss with others and try out your ideas.

> Suggestion: if you want to test your alert and "make your application" unhealthy you can remove the deployment from kubernetes or modify the healthcheck code to return 500 and redeploy. 

<details>
<summary>See the ideas!</summary>

> Some ideas (there are probably even more options to do it):

Author: Fixed

2. [**Alerts for Specific Exceptions with Application Insights**](https://stackoverflow.com/questions/47147651/creating-alerts-for-specific-exceptions-with-application-insight-microsoft-azur)
3. Specific for AKS: Resource logs, [Container insights log alert]([Log alerts from Container insights - Azure Monitor | Microsoft Learn](https://learn.microsoft.com/en-us/azure/azure-monitor/containers/container-insights-log-alerts))
4. [Availability Alerts with Application Insights](https://learn.microsoft.com/en-us/azure/azure-monitor/app/availability-alerts)

Let's take a closer look at the last option!

Go to Application Insights -> Availability section. Now create a standard test posting an endpoint to your application health check:

> How to find your application IP?
>
> ```shell
> DEVICES_API_IP=$(kubectl get service devices-api-service -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
> HEALTHCHECK_URL="http://$DEVICES_API_IP:8080/health"
> ```

![](.attachments/availability_standard_test.png)

The alert will be automatically created for you!

![](.attachments/availability_alert.png)

</details>

> ## [Optional] Infrastructure as Code

Reviewer 2: General question. Great we have this for the alerts, but do we want leave this out for the dry run, till we have the full IaC also for the dashboard and alerts section? Just to stay consistent?

Author: I would leave it as it's marked as optional, so it's something extra - e.g. for people who finished earlier?
